ed_sensor_integration/tools/fitter_viz.cpp

```diff
@@ -0,0 +1,382 @@
+
+#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+
+#include <tue/config/read.h>
+#include <tue/config/reader.h>
+#include <tue/config/data_pointer.h>
+
+#include <ed/entity.h>
+
+#include "geolib/math_types.h"
+#include <tf2/convert.h>
+#include <tf2/utils.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+
+#include <fstream>
+#include <vector>
+
+// ----------------------------------------------------------------------------------------------------
+
+struct Snapshot
+{
+    rgbd::ImagePtr image;
+    geo::Pose3D sensor_pose;
+    std::string area_description;
+    ed::WorldModel world_model;
+};
+
+// ----------------------------------------------------------------------------------------------------
+
+bool readImage(const std::string& filename, rgbd::ImagePtr& image, geo::Pose3D& sensor_pose,
+               ed::WorldModel& world_model, std::string& area_description)
+{
+    tue::config::DataPointer meta_data;
+
+    try
+    {
+        meta_data = tue::config::fromFile(filename);
+    }
+    catch (tue::config::ParseException& e)
+    {
+        std::cerr << "Could not open '" << filename << "'.\n\n" << e.what() << std::endl;
+        return false;
+    }
+
+    tue::config::Reader r(meta_data);
+
+    // Read image
+    std::string rgbd_filename;
+    if (r.value("rgbd_filename", rgbd_filename))
+    {
+        tue::filesystem::Path abs_rgbd_filename = tue::filesystem::Path(filename).parentPath().join(rgbd_filename);
+
+        std::ifstream f_rgbd;
+        f_rgbd.open(abs_rgbd_filename.string().c_str(), std::ifstream::binary);
+
+        if (!f_rgbd.is_open())
+        {
+            std::cerr << "Could not open '" << filename << "'." << std::endl;
+            return false;
+        }
+
+        image.reset(new rgbd::Image);
+
+        tue::serialization::InputArchive a_in(f_rgbd);
+        rgbd::deserialize(a_in, *image);
+    }
+
+    // Read sensor pose
+    if (!ed::deserialize(r, "sensor_pose", sensor_pose))
+    {
+        std::cerr << "No field 'sensor_pose' specified." << std::endl;
+        return false;
+    }
+
+//    if (r.hasError())
+//    {
+//        std::cout << "Error while reading file '" << filename << "':\n\n" << r.error() << std::endl;
+//        return false;
+//    }
+
+    return true;
+}
+
+// ----------------------------------------------------------------------------------------------------
+
+bool loadWorldModel(const std::string& model_name, ed::WorldModel& world_model)
+{
+    ed::UpdateRequest req;
+
+    ed::models::ModelLoader model_loader;
+
+    std::stringstream error;
+    if (!model_loader.create("_root", model_name, req, error, true))
+    {
+        std::cerr << "Model '" << model_name << "' could not be loaded:" << std::endl << std::endl;
+        std::cerr << error.str() << std::endl;
+        return false;
+    }
+
+    // Reset world
+    world_model = ed::WorldModel();
+
+    // Update world
+    world_model.update(req);
+
+    return true;
+}
```

Comment:
I think it is better to return a world model pointer. In case it is a nullptr it is valid.

Because now you are creating a WM, pass it by reference to this function. Then you create a new WM and update it. The first instance is useless.